fs/fs_zenfs.cc

```diff
@@ -477,40 +482,39 @@ std::shared_ptr<ZoneFile> ZenFS::GetFileInternal(std::string fname) {
 std::shared_ptr<ZoneFile> ZenFS::GetFile(std::string fname) {
   std::shared_ptr<ZoneFile> zoneFile(nullptr);
   std::lock_guard<std::mutex> lock(files_mtx_);
-  zoneFile = GetFileInternal(fname);
+  zoneFile = GetFileNoLock(fname);
   return zoneFile;
 }
 
-IOStatus ZenFS::DeleteFile(std::string fname) {
+/* Must hold files_mtx_ */
+IOStatus ZenFS::DeleteFileNoLock(std::string fname, const IOOptions& options,
+                                 IODebugContext* dbg) {
   std::shared_ptr<ZoneFile> zoneFile(nullptr);
   IOStatus s;
 
-  {
-    std::lock_guard<std::mutex> lock(files_mtx_);
-    zoneFile = GetFileInternal(fname);
-    if (zoneFile != nullptr) {
-      std::string record;
-
-      zoneFile = files_[fname];
-      files_.erase(fname);
+  zoneFile = GetFileNoLock(fname);
+  if (zoneFile != nullptr) {
+    std::string record;
 
-      EncodeFileDeletionTo(zoneFile, &record);
-      s = PersistRecord(record);
-      if (!s.ok()) {
-        /* Failed to persist the delete, return to a consistent state */
-        files_.insert(std::make_pair(fname.c_str(), zoneFile));
-      } else {
-        /* Mark up the file as deleted so it won't be migrated by GC */
-        zoneFile->SetDeleted();
-        zoneFile.reset();
-      }
+    if (zoneFile->IsOpenForWR())
+      return IOStatus::Busy("ZenFS::DeleteFileNoLock(): file open for writing:",
+                            fname.c_str());
+    zoneFile = files_[fname];
+    files_.erase(fname);
+    EncodeFileDeletionTo(zoneFile, &record);
+    s = PersistRecord(record);
+    if (!s.ok()) {
+      /* Failed to persist the delete, return to a consistent state */
+      files_.insert(std::make_pair(fname.c_str(), zoneFile));
     } else {
-      s = IOStatus::NotFound("ZenFS::DeleteFile(): File not found");
+      /* Mark up the file as deleted so it won't be migrated by GC */
+      zoneFile->SetDeleted();
+      zoneFile.reset();
     }
+  } else {
+    s = target()->DeleteFile(ToAuxPath(fname), options, dbg);
```

Comment:
I don't think that we need to hold `files_mtx_` across deletion on aux fs. I does not break anything, but not strictly needed.

Reply:
Yeah, I thought about this, if we specifically want to keep it out of the files_mtx_ we would need to introduce one more function and more plumbing is involved, so this seemed better than the alternative.
A recursive locking mechanism can be helpful to reduce these extra plumbing.